### ISSUE TITLE
Modernize and improve vimgolf.com site

### ISSUE THREAD

**Issue by igrigorik** [igrigorik/vimgolf#171]
First, a little history.. VimGolf was born as a fun holiday project back in 2010 and has far surpassed my expectation in terms of engagement, longevity, and community. Over the past six years it attracted 17K+ users, 170K+ submissions, 340+ user submitted challenges, and a dozen super active moderators (most of the credit for all of the success goes to them!) who have made it all work.

Code: https://github.com/igrigorik/vimgolf/tree/web

Six years later, the site infrastructure needs some love and attention:
1. It currently runs on a dated Rails 3.0.19 stack
1. Many of the dependencies are also dated and need to be updated
1. The data model needs to be rethought to fix performance issues and enable new functionality

Original assumption was dozens of challenges with tens~hundreds of entries. Leaderboard and top user functionality was bolted on later and not very smartly at that -- it worked fine at the time, but needs to be redone to scale further. In terms of the path forward:

- Ideally stick with Ruby+Rails, as it allows us to easily integrate and share vimgolf CLI code
- Data storage can be migrated to whatever fits the new requirements; we're not tied to MongoDB

**Interested in helping out?** Ping this thread and [let's collaborate](https://www.igvita.com/2011/12/19/dont-push-your-pull-requests/).

**Comment by stevecrozz** [comment]
I wouldn't mind helping. Are we going to organize a plan right here in this thread, or are you just polling for volunteers at the moment?

**Comment by davidCarlos** [comment]
I would like to help.

**Comment by nelsonsar** [comment]
I'm not an expert but, if I can help please let me know :)

**Comment by tomciopp** [comment]
@igrigorik Would love to help out. Can you give any context with regard to where perf issues are occurring? What new features would you like to develop?

**Comment by catilac** [comment]
I haven't done Rails in a while, but I used to be an expert. I'm happy to help

**Comment by igrigorik** [comment]
\o/ ... looking forward to working with you guys!

First off, I'm open to any ideas and suggestions for how to coordinate this work between everyone. Let me know if you have any preferences. In the meantime, I'll try to sketch out a vague outline of what I had in mind: 

- We **definitely** need to move a modern Rails stack to get the benefits of modern APIs, functionality, security updates, and to make it more accessible to work on in general. One of the problems right now is that Mongoid and other dependencies have revamped their APIs, so it's not as simple as `bundle update`, and this also prevents us from using latest and greatest features...
- Performance problems: if you look at the data model, you basically have two structures (users and challenges). Challenges embeds 'entries', and entries have 'comments'.
 - The original assumption was that anytime you fetch a challenge you want to fetch all the entries. This is still mostly true, but some challenges have grown very large (5K+ entries!), so this is causing trouble -- we end up pulling multi megabyte responses from mongo. 
 - The next problem is user scores + leaderboard aggregates. I disabled those paths for time being because they were locking up our unicorn instances: in order to compute user's page we'd have to fetch each challenge they partcipated in, plus all the entries for each, and find position of the user in the entry list; for leaderboard things get even more expensive because we have to run the same reduce over all challenges and compute everyone's rank.

None of the above is particularly exceptional or hard, and the amount of data is not *that big*, but the current structure is nonethless not capable of dealing with this too well.. Unless someone can point out some mongo tricks to make all this nice and fast.

I guess one way to simulate the problem would be to generate a test set: ~1K challenges, 300K entries, 50K users.. or some such. You'll immediately spot the bottlenecks. ;-)

----

All that aside. I'm wondering if a good starting point would be to keep the data structures as is, and start by upgrading to modern Rails + Mongoid version. With that in place, we'll be in a good place to start iterating on updating the data abstractions and other features?

**Comment by tomciopp** [comment]
@igrigorik What are your thoughts on converting the app to a relational db? (Postgresql) Based on a quick overview of the data model nothing really jumps out as a justification for using a NoSQL solution over a relational db. (it would also likely be more accessible to most rails devs) Also there aren't many tables so we shouldn't see too much pain in a conversion. The other things that jump out are breaking the web app and gem into separate repos, decoupling authorization from the models, adding proper CI, and simplifying the controllers. 

Some additional questions:
- What are your thoughts on updates to the front end? 
- Do we want to add additional methods for authentication? (Devise, Github, etc.)

**Comment by catilac** [comment]
I agree with the starting point of upgrading Rails + Mongoid.
Additionally, I agree with @tomciopp on all counts.

Would we be interested in creating a modern client side? Using React or Vue? Or do we just want to focus on upgrading the tools and fixing performance issues?

**Comment by igrigorik** [comment]
> @igrigorik What are your thoughts on converting the app to a relational db? (Postgresql) Based on a quick overview of the data model nothing really jumps out as a justification for using a NoSQL solution over a relational db.

If memory serves, the only reason I went with MongoDB at the time is because I was fishing for a reason to play with Mongoid -- I know, I know. I'm good with switching to Postgres, or anything else, if it makes sense. On that note, I guess the relevant question is: what are the queries we need to fulfill? 

- Home page: list of challenges with entry counts, sorted by ~interestingness. 
- Challenge page: 
   - List of entries sorted by {score, date}
      -- Each entry has a list of comments
      -- User can have many submissions per challenge but we only show highest ranking
   - Some "load more" / pagination function to avoid loading thousand of entries
   - Efficient way to filter out higher ranking (lower score) submissions, but allow to see a few above
- User page:
  - Global rank of the user
  - List of contributed challenges
  - List of participated challenges
  - For each of the above: rank in the challenge
- Leaderboard
  - Sorted list of users by rank

My hunch is that we'll probably need to compute some of the ranks whenever the user submits an entry.. Right now, we do it on demand, and its a rather expensive operation. That said, I'd also like to keep this ~real-time.

> What are your thoughts on updates to the front end?

You mean the CSS/JS? All for it, but low priority relative to other things.

> Do we want to add additional methods for authentication? (Devise, Github, etc.)

No objections, as long as we keep the signup/sign-in flow simple.

**Comment by nelsonsar** [comment]
> Right now, we do it on demand, and its a rather expensive operation. That said, I'd also like to keep this ~real-time.

We can maybe use PostgreSQL materialized views to achieve this.

Anyway, will you create tasks for Rails upgrades? Should we do it version by version following [this](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-3-0-to-rails-3-1)? The test coverage is enough? Maybe is a good idea adding more tests before the upgrades (if it's the case for it, of course).

**Comment by tomciopp** [comment]
@igrigorik @nelsonsar It might be smart to introduce a handicapping system similar to how real golf evaluates player rating. They have the exact same problem of evaluating skill level based on performing similar tasks in different scenarios. Calculate someone's performance vs the baseline on all challenges => compute and cache score once challenge is finished => compare scores between all users.

**Comment by zoso10** [comment]
What are people's thoughts on adding some integration tests? Did a little poking and there's some basic specs around a few controller endpoints. The current suite is with `rspec` so we could get up and running with `capybara` easily. IMO it would be a lot safer to have better test coverage before making a lot of changes, especially for upgrading. 

Also did a quick peek at the current gem list and there's not many dependencies on specific rails gems (`activesupport`, `actionpack`, etc) versions so we could mostly handle deprecations within rails framework, bumping other gems, and upgrading mongo/db independently.

**Comment by aaooki** [comment]
I'd love to help

**Comment by DanAndreasson** [comment]
I'd love to chip in!

**Comment by hvenables** [comment]
I'd love to help out

**Comment by igrigorik** [comment]
@zoso10 @nelsonsar I'm all for adding more tests. I'd also love to see a CI setup with Travis, so we can get feedback on any pull requests and updates.

@tomciopp good points on ranking. This is probably worth a separate discussion thread? You can take a look at how we currently rank users in the users model/controller.

In terms of tasks, I think I see some themes emerging here:

1. Improve test coverage and infrastructure - issue https://github.com/igrigorik/vimgolf/issues/172.
1. Upgrade Rails and dependencies to current version - issue https://github.com/igrigorik/vimgolf/issues/173.
1. Investigate new data model - https://github.com/igrigorik/vimgolf/issues/174.
1. Update the front-end - https://github.com/igrigorik/vimgolf/issues/175.

What else is missing? Let's drill into each of the above in the individual threads. I'm happy to add folks as collaborators on the repo to help streamline the upgrade / review process.

Last but not least, some thoughts on repo structure: https://github.com/igrigorik/vimgolf/issues/176.

**Comment by sunjay** [comment]
What do you think of managing all of these tasks using GitHub's projects feature? It's better integrated than a Trello board and works quite well. 

https://github.com/igrigorik/vimgolf/projects

**Comment by sunjay** [comment]
If time permits I'm happy to contribute to the frontend. Never used rails.

**Comment by igrigorik** [comment]
@sunjay yep, done: https://github.com/igrigorik/vimgolf/projects/1

**Comment by igrigorik** [comment]
Resolving. Running modern rails, and we landed some data model fixes (https://github.com/igrigorik/vimgolf/pull/248).

Thanks all! :-)
